### PR TITLE
Add Support for requests with no response

### PR DESCRIPTION
**Pull request by jshaughn**
Requests that have no response must also be treated as errors because the request never completes, and a response is never sent.  Error handling up to this point assumed a valid response code for the request protocol. The graph config "responses" can now pass code="-" for http and grpc to indicate the no response case.  And each protocol will include "noResponse" traffic rates.
This adds the side panel add support for the new error case.

Also:
- Fix bug in ErrorRate.sumRequests where inbound rates could be corrupted with outbound rates.
- Fix minor regression with health badge in side panel by restoring the 'health' text

**Requires Server PR: https://github.com/kiali/kiali/pull/3214**

Part of https://github.com/kiali/kiali/issues/3194

For testing notes see the server PR.

@jmazzitelli please review the graph code changes
@jotak please review the metrics code changes
@aljesusg please review the health code changes

### CODE REVIEW

File: src/components/Health/__tests__/HealthIndicator.test.tsx [hunk 1]
```diff
@@ -140,6 +140,6 @@ describe('HealthIndicator', () => {
     html = wrapper.html();
     expect(html).toContain(PFAlertColor.Danger);
     expect(html).toContain('Outbound: 16.67%');
-    expect(html).toContain('Inbound: 26.47%');
+    expect(html).toContain('Inbound: 50.00%');
```
Comment (jotak) [inline src/components/Health/__tests__/HealthIndicator.test.tsx:143]: hmm .. so there was a bug here?

File: src/components/SummaryPanel/RateChart.tsx [hunk 2]
```diff
@@ -200,15 +209,23 @@ export const renderInOutRateChartHttp = (
...
       color: colorVals.ChartDanger
-    }
+    },
+    {
+      name: 'No Response',
+      dp: [
+        { x: 'In', y: percentNRIn },
+        { x: 'Out', y: percentNROut }
+      ],
+      color: colorVals.ChartDanger
```
Comment (jotak) [inline src/components/SummaryPanel/RateChart.tsx:227]: that's not black?

File: src/types/ErrorRate/ErrorRate.ts [hunk 3]
```diff
@@ -3,9 +3,16 @@ import { RateHealthConfig, ToleranceConfig } from '../ServerConfig';
...
 export const sumRequests = (inbound: RequestType, outbound: RequestType): RequestType => {
-  let result: RequestType = inbound;
+  let result: RequestType = {};
+  // init result with a deep clone of inbound
+  for (let [protocol, req] of Object.entries(inbound)) {
+    result[protocol] = {};
+    for (let [code, rate] of Object.entries(req)) {
+      result[protocol][code] = rate;
+    }
+  }
```
Comment (aljesusg) [inline src/types/ErrorRate/ErrorRate.ts:15]: Good one, I didn't realize this :(

**Review by aljesusg — APPROVED** [review]
LGFM from my side erroRate

File: src/components/Health/__tests__/HealthIndicator.test.tsx (hunk 1 above)
Comment (jshaughn) [reply src/components/Health/__tests__/HealthIndicator.test.tsx:143]: @jotak Right, the bug fix in ErrorRate.ts uncovered the bad expectation in this test.  So, a bug in the code and a bug in the test, a bad combination but it happens.

File: src/types/ErrorRate/ErrorRate.ts (hunk 3 above)
Comment (jshaughn) [reply src/types/ErrorRate/ErrorRate.ts:15]: @aljesusg This was a tricky one to figure out.  But yes, the by-reference assignment led to corruption of the inbound rates.

File: src/components/SummaryPanel/RateChart.tsx (hunk 2 above)
Comment (jshaughn) [reply src/components/SummaryPanel/RateChart.tsx:227]: Good catch, @jotak !

**Review by jmazzitelli — APPROVED** [review]

**Review by jotak — APPROVED** [review]
LGTM for my part. I smoke-tested just some non-regression (I wasn't able to trigger 0-code responses .. even with injected delay, not sure why)

**Comment by pbajjuri20** [comment]
I have tested yesterday, Just waited for Dev-Approval. I don't find any issues LGTM.



flags


health after

**Pull request merged**